components/ai_chat/core/browser/ai_chat_service.cc

```diff
@@ -231,6 +358,406 @@ void AIChatService::OnMemoryEnabledChanged() {
+
+  DVLOG(1) << "Ollama sync preference changed to: " << ollama_sync_enabled;
+
+  if (ollama_sync_enabled) {
+    DVLOG(1) << "Ollama sync enabled - syncing models";
+    SyncOllamaModels();
+  } else {
+    DVLOG(1) << "Ollama sync disabled - will remove Ollama models";
+    RemoveOllamaModels();
+  }
+}
+
+void AIChatService::RemoveOllamaModels() {
+  if (!model_service_) {
+    return;
+  }
+
+  const auto& all_models = model_service_->GetModels();
+  std::vector<uint32_t> ollama_custom_indices; // Indices within custom models only
+
+  DVLOG(1) << "Total models in service: " << all_models.size();
+
+  // Find custom models and identify which ones are Ollama models
+  uint32_t custom_model_index = 0;
+  for (size_t i = 0; i < all_models.size(); ++i) {
+    const auto& model = all_models[i];
+    DVLOG(1) << "Model at index " << i << ": " << model->display_name;
+
+    if (model->options && model->options->is_custom_model_options()) {
+      // This is a custom model
+      DVLOG(1) << "Custom model at full-index " << i << ", custom-index " << custom_model_index << ": " << model->display_name;
+
+      if (model->options->get_custom_model_options() &&
+          model->options->get_custom_model_options()->endpoint.is_valid() &&
+          model->options->get_custom_model_options()->endpoint.spec() ==
+          "http://localhost:11434/v1/chat/completions") {
+        DVLOG(1) << "Found Ollama model at custom-index " << custom_model_index << ": " << model->display_name;
+        ollama_custom_indices.push_back(custom_model_index);
+      }
+      custom_model_index++;
+    }
+  }
+
+  // Remove Ollama models (in reverse order to maintain indices)
+  for (auto it = ollama_custom_indices.rbegin(); it != ollama_custom_indices.rend(); ++it) {
+    DVLOG(1) << "Removing Ollama model at custom-index: " << *it;
+    model_service_->DeleteCustomModel(*it);
+    DVLOG(1) << "Called DeleteCustomModel(" << *it << ")";
+  }
+
+  if (!ollama_custom_indices.empty()) {
+    DVLOG(1) << "Removed " << ollama_custom_indices.size() << " Ollama models";
+
+    // Verify removal by checking if any Ollama models still exist
+    const auto& updated_models = model_service_->GetModels();
+    int remaining_ollama_count = 0;
+    for (const auto& model : updated_models) {
+      if (model->options && model->options->is_custom_model_options() &&
+          model->options->get_custom_model_options() &&
+          model->options->get_custom_model_options()->endpoint.is_valid() &&
+          model->options->get_custom_model_options()->endpoint.spec() ==
+          "http://localhost:11434/v1/chat/completions") {
+        remaining_ollama_count++;
+        DVLOG(1) << "Still found Ollama model after removal: " << model->options->get_custom_model_options()->model_request_name;
+      }
+    }
+    DVLOG(1) << "Remaining Ollama models after removal: " << remaining_ollama_count;
+  }
+}
+
+void AIChatService::SyncOllamaModels() {
```

Comment:
My initial impression is that this should all be handled by the `ai_chat::ModelService` and a lot of it put in to a separate ollama client class so that it's not taking up hundreds of lines there.